control-plane-operator/hostedclusterconfigoperator/controllers/machine/machine.go

```diff
@@ -180,3 +219,62 @@ func serviceHasIPFamily(service *corev1.Service, ipFamilyToFind corev1.IPFamily)
 	}
 	return false
 }
+
+func kubevirtInfraNamespace(hcp *hyperv1.HostedControlPlane) string {
+	if hcp.Spec.Platform.Kubevirt.Credentials != nil {
+		return hcp.Spec.Platform.Kubevirt.Credentials.InfraNamespace
+	}
+	return hcp.Namespace
+}
+
+func (r *reconciler) removeKubevirtPassthrowEndpointSlices(ctx context.Context, svc *corev1.Service) error {
+	if err := r.kubevirtInfraClient.DeleteAllOf(ctx, &discoveryv1.EndpointSlice{}, client.MatchingLabels{
+		discoveryv1.LabelServiceName: svc.Name,
+	}); err != nil {
+		return err
+	}
+	return r.removeServiceFinalizer(ctx, svc)
+}
+
+func (r *reconciler) removeServiceFinalizer(ctx context.Context, svc *corev1.Service) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(svc), svc); err != nil {
+			return err
+		}
+		if controllerutil.ContainsFinalizer(svc, serviceFinalizerName) {
+			if controllerutil.RemoveFinalizer(svc, serviceFinalizerName) {
+				return r.kubevirtInfraClient.Update(ctx, svc)
+			}
+		}
+		return nil
+	})
+}
+
+func (r *reconciler) addServiceFinalizer(ctx context.Context, svc *corev1.Service) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(svc), svc); err != nil {
+			return err
+		}
+		if !controllerutil.ContainsFinalizer(svc, serviceFinalizerName) {
+			if controllerutil.AddFinalizer(svc, serviceFinalizerName) {
+				return r.kubevirtInfraClient.Update(ctx, svc)
+			}
```

Comment:
can Patch() be used instead of Update here for the finalizer addition/removal?

Reply:
I have follow kubebuilder guidelines since this is just reconciled by us is not going to conflict that much or at all https://book.kubebuilder.io/reference/using-finalizers#using-finalizers

Reply:
Changed to use Patch and removed retry on conflict.